in-progress/6973-refactor-base-contract-interaction.md

```diff
@@ -0,0 +1,1092 @@
+}
+
+export interface UserFunctionCall {
+  contractInstance: ContractInstanceWithAddress;
+  functionName: string;
+  args: any;
+  deploymentOptions?: DeploymentOptions;
+  contractArtifact?: ContractArtifact;
+  functionAbi?: FunctionAbi;
+}
+
+// new
+export interface UserRequest {
+  calls: UserFunctionCall[];
+  gasSettings?: GasSettings;
+  paymentMethod?: FeePaymentMethod;
+  from?: AztecAddress;
+  simulatePublicFunctions?: boolean;
+  executionResult?: ExecutionResult; // the raw output of a simulation that can be proven
+  tx?: Tx; // a proven tx to send
+}
+
+export type TxExecutionRequestAdapter = (builder: TxExecutionRequestBuilder, userRequest: UserRequest) => Promise<void>;
+
+export interface TxExecutionRequestComponent {
+  adaptTxExecutionRequest: TxExecutionRequestAdapter;
+}
+
+export interface FeePaymentMethod extends TxExecutionRequestComponent {
+  getEquivalentAztBalance(): Promise<Fr>;
+}
+
+export interface SimulationOutput {
+  tx: Tx;
+  result: DecodedReturn | [];
+  request: UserRequest;
+  executionResult: ExecutionResult;
+  publicOutput: PublicSimulationOutput;
+  privateOutput: NestedProcessReturnValues;
+  error?: SimulationError;
+}
+
+export interface UserAPI {
+  getTxExecutionRequest(userRequest: UserRequest): Promise<TxExecutionRequest>;
+  simulate(userRequest: UserRequest): Promise<SimulationOutput>;
+  read(userRequest: UserRequest): Promise<SimulationOutput>;
+  prove(userRequest: UserRequest): Promise<UserRequest>;
+  send(userRequest: UserRequest): Promise<SentTx>;
+}
+
+export type Wallet = UserAPI & AccountInterface & PXE & AccountKeyRotationInterface;
+
+// Swap `EntrypointInterface` for `TxExecutionRequestComponent`
+export interface AccountInterface extends AuthWitnessProvider, TxExecutionRequestComponent {
+  getCompleteAddress(): CompleteAddress;
+  getAddress(): AztecAddress;
+  getChainId(): Fr;
+  getVersion(): Fr;
+
+}
+
+// unchanged
+export interface AuthWitnessProvider {
+  createAuthWit(
+    messageHashOrIntent:
+      | Fr
+      | Buffer
+      | {
+          caller: AztecAddress;
+          action: ContractFunctionInteraction | FunctionCall;
+          chainId?: Fr;
+          version?: Fr;
+        },
+  ): Promise<AuthWitness>;
+}
+
+```
+
+
+## Implementation
+
+### Expose an explicit `proveTx` on PXE
+
+We currently don't have a way to *only* prove a transaction; everything must be simulated first.
+
+So we will expose a `proveTx` method on the PXE that will take a `TxExecutionRequest` and an `ExecutionResult` and return a `Tx`.
+
+### Drop `isFeePayer` from `FeeEntrypointPayload`
+
+
+Removing this makes the code simpler and more flexible: account contracts can just check if fee payer has been set and set it if not.
+
+Concerned users can just inspect the simulation result to see if they are paying the fee.
+
+Then we just have `EntrypointPayload` as a single, concrete class.
+
+### BaseWallet.getTxExecutionRequest
+
+Consider that we have, e.g.:
+
+```ts
+{
+  calls: [{
```

Comment:
While this interface might make it more clear, for cases where it is not a multi-call it seems worse to me, seems more tricky to get nice autocomplete for and early compiler error or type errors?

Reply:
I couldn't come up with a way to support multi-call and single call without accepting an array here or doubling the size of the `UserAPI` to have, e.g. `simulateMultiCall`. 

I'm kind of thinking that the typescript that gets generated for each contract can actually still help us a lot, and provide autocomplete etc, but they will just return these (much simpler) `UserFunctionRequests`. Then the fact that `calls` is an array is maybe not so bad.

E.g.

```ts
{
  calls: [bananaCoinInstance.transfer({ 
      from: aliceAddress,
      to: bobAddress,
      value: privateBalance,
      nonce: 0n
    })]
}
```

My hope with this design is to just some solid low-level primitives that we can build more ergonomics on top of after they're a bit more battle tested. So I don't mind forgoing autocomplete in the immediate term if we think we can extend this (or some other) design to support it.